sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -739,13 +740,23 @@ class Dataset[T] private[sql](
   // We only accept an existing column name, not a derived column here as a watermark that is
   // defined on a derived column cannot referenced elsewhere in the plan.
   def withWatermark(eventTime: String, delayThreshold: String): Dataset[T] = withTypedPlan {
-    val parsedDelay =
-      try {
+    val parsedDelay = try {
+      if (delayThreshold.toLowerCase(Locale.ROOT).startsWith("interval")) {
+        CatalystSqlParser.parseExpression(delayThreshold) match {
+          case Literal(months: Int, _: YearMonthIntervalType) =>
+            new CalendarInterval(months, 0, 0)
+          case Literal(micros: Long, _: DayTimeIntervalType) =>
+            val days = micros / DateTimeConstants.MICROS_PER_DAY
+            val restMicros = micros % DateTimeConstants.MICROS_PER_DAY
```

Comment:
What's the reason to perform such normalization? Why cannot you just pass `micros` as the third parameter?

Reply:
Ah, we don't need to normalize. Thanks.